### PR TITLE
#2277 - Overapproximate a polytope with a zonotope using cartesian decomposition

### LINKED ISSUE #2277: Overapproximate a polytope with a zonotope using cartesian decomposition
see sec 8.2.4 [here](https://tel.archives-ouvertes.fr/tel-00422569v2/document)

### PR DESCRIPTION
**Pull request by SebastianGuadalupe**
Closes #2277.

### CODE REVIEW

File: src/Approximations/overapproximate.jl [hunk 1]
```diff
@@ -1636,6 +1647,31 @@ function overapproximate(X::LazySet{N}, ::Type{<:Zonotope},
     return Zonotope(c, G)
 end
 
+function _overapproximate_zonotope_cpa(X::AbstractPolytope, ::Type{<:Zonotope},
+                                       dirs=OctDirections)
+    n = dim(X)
+    p = Vector(undef, Int(floor(n/2)))
+    j = 1
+    if isa(X, VPolytope)
```
Comment (mforets) [inline src/Approximations/overapproximate.jl:1655]: why not dispatch on `X`? you can use a helper function to reduce the repeated code.

File: src/Approximations/overapproximate.jl [hunk 2]
```diff
@@ -1636,6 +1647,31 @@ function overapproximate(X::LazySet{N}, ::Type{<:Zonotope},
...
+        for i=1:2:n-1
+            p[j] = project(X, [i, i+1])
+            j += 1
+        end
+    else
+        for i=1:2:n-1
+            p[j] = concretize(Projection(X, [i, i+1]))
```
Comment (mforets) [inline src/Approximations/overapproximate.jl:1662]: how does this differ from `project(X, [i, i+1])` ?

File: src/Approximations/overapproximate.jl [hunk 3]
```diff
@@ -1636,6 +1647,31 @@ function overapproximate(X::LazySet{N}, ::Type{<:Zonotope},
     return Zonotope(c, G)
 end
 
+function _overapproximate_zonotope_cpa(X::AbstractPolytope, ::Type{<:Zonotope},
+                                       dirs=OctDirections)
+    n = dim(X)
+    p = Vector(undef, Int(floor(n/2)))
```
Comment (mforets) [inline src/Approximations/overapproximate.jl:1653]: can we concretely type `p`?

File: src/Approximations/overapproximate.jl [hunk 4]
```diff
@@ -1636,6 +1647,31 @@ function overapproximate(X::LazySet{N}, ::Type{<:Zonotope},
...
+            p[j] = concretize(Projection(X, [i, i+1]))
+            j += 1
+        end
+    end
+
+    if !iseven(n)
+        push!(p, project(X, [n]))
+    end
+
+    Z = [_overapproximate_zonotope_vrep(poly, Zonotope, dirs(2)) for poly in p]
```
Comment (mforets) [inline src/Approximations/overapproximate.jl:1671]: if `n` is not even, the last block has dimension 1. thn in that case, using `dirs(2)` is wrong?

File: src/Approximations/overapproximate.jl (hunk 3 above)
Comment (mforets) [reply src/Approximations/overapproximate.jl:1653]: .. if `project` uses dynamic dispatch (VPolytpe/VPolygon/Interval) then it gets more complicated. that's the reason that i find it useful to have type-stable internal functions as suggested [here](https://github.com/JuliaReach/LazySets.jl/pull/2287#issuecomment-665160969).

Comment (SebastianGuadalupe) [reply src/Approximations/overapproximate.jl:1653]: Yes, that is the problem, when you project a `VPolytope`, it gives you a `VPolygon`

Comment (mforets) [reply src/Approximations/overapproximate.jl:1653]: hmm ok but the blocks are all of size 2 here. then we can use `p` of type VPolygon if the system dim is even and a union of VPolygon and interval if it's odd.

Comment (SebastianGuadalupe) [reply src/Approximations/overapproximate.jl:1653]: That's a good idea! I will do that.

**Comment by mforets** [comment]
remains that we add some tests, and that we have https://github.com/JuliaReach/LazySets.jl/issues/2299.

File: src/Approximations/overapproximate.jl [hunk 5]
```diff
@@ -1636,6 +1687,48 @@ function overapproximate(X::LazySet{N}, ::Type{<:Zonotope},
...
 
+# overload on direction type
+function _overapproximate_zonotope_vrep(X::LazySet{N},
+                                        dir::Type{<:AbstractDirections{N}};
+                                        solver=default_lp_solver(N)) where {N}
+    n = dim(X)
+    return _overapproximate_zonotope_vrep(X, dir(2), solver=solver)
+end
+
+function _overapproximate_zonotope_cpa(X::LazySet, dir::Type{<:AbstractDirections})
```
Comment (mforets) [inline src/Approximations/overapproximate.jl:1698]: the docstring for this one is also missing @SebastianGuadalupe

**Comment by mforets** [comment]
looks good so far. @SebastianGuadalupe can you rebase this branch and add some tests for the new function?

File: docs/src/lib/sets/Zonotope.md [hunk 6]
```diff
@@ -18,7 +18,9 @@ split(::AbstractZonotope, ::Int)
 split(::AbstractZonotope, ::AbstractVector{Int}, ::AbstractVector{Int})
 remove_zero_generators(::Zonotope)
 linear_map!(::Zonotope, ::AbstractMatrix, ::Zonotope)
-quadratic_map(Q::Vector{MT}, Z::Zonotope{N}) where {N, MT<:AbstractMatrix{N}}
+quadratic_map(::Vector{MT}, ::Zonotope{N}) where {N, MT<:AbstractMatrix{N}}
+LazySets._overapproximate_zonotope_vrep
+LazySets._overapproximate_zonotope_cpa
```
Comment (schillic) [inline docs/src/lib/sets/Zonotope.md:23]: These must go to `approximations.md` and need `Approximations` as prefix instead.

File: src/Approximations/overapproximate.jl [hunk 7]
```diff
@@ -1514,17 +1514,68 @@ function overapproximate(am::AbstractAffineMap{N, <:AbstractHyperrectangle{N}},
...
+### Output
+
+A zonotope that overapproximates `X` and uses at most the directions provided in
+`dir` (redundant directions will be ignored).
+
+### Notes
+
+Two algorithms are available:
+
+- `"vrep"` -- Overapproximate a polytopic set with a zonotope.of minimal total generator sum
```
Comment (schillic) [inline src/Approximations/overapproximate.jl:1540]:
```suggestion
- `"vrep"` -- Overapproximate a polytopic set with a zonotope of minimal total generator sum
```

File: src/Approximations/overapproximate.jl [hunk 8]
```diff
@@ -1636,6 +1687,74 @@ function overapproximate(X::LazySet{N}, ::Type{<:Zonotope},
     return Zonotope(c, G)
 end
 
+# overload on direction type
+function _overapproximate_zonotope_vrep(X::LazySet{N},
+                                        dir::Type{<:AbstractDirections{N}};
+                                        solver=default_lp_solver(N)) where {N}
+    n = dim(X)
+    return _overapproximate_zonotope_vrep(X, dir(2), solver=solver)
```
Comment (schillic) [inline src/Approximations/overapproximate.jl:1695]:
```suggestion
    return _overapproximate_zonotope_vrep(X, dir(dim(X)), solver=solver)
```

File: src/Approximations/overapproximate.jl [hunk 9]
```diff
@@ -1636,6 +1687,74 @@ function overapproximate(X::LazySet{N}, ::Type{<:Zonotope},
...
+- `dir`      -- directions used for the generators
+
+### Output
+
+A zonotope that overapproximates `X`.
+
+### Notes
+
+The algorithm decomposes `X` in 2D sets and overapproximates those sets with
+zonotopes, then does a cartesian product with the sets.
```
Comment (schillic) [inline src/Approximations/overapproximate.jl:1715]:
```suggestion
zonotopes, and finally takes cartesian product of the sets and converts to a zonotope.
```

File: src/Approximations/overapproximate.jl [hunk 10]
```diff
@@ -1636,6 +1687,74 @@ function overapproximate(X::LazySet{N}, ::Type{<:Zonotope},
...
+
+[1] Le Guernic, C. (2009). Reachability analysis of hybrid systems with linear
+continuous dynamics (Doctoral dissertation).
+"""
+function _overapproximate_zonotope_cpa(X::LazySet, dir::Type{<:AbstractDirections})
+    n = dim(X)
+    nblocks = Int(floor(n/2))
+
+    # overapproximate 2D blocks
+    if n > 1
```
Comment (schillic) [inline src/Approximations/overapproximate.jl:1729]:
```suggestion

    # overapproximate 2D blocks
    if n > 1
        nblocks = Int(floor(n/2))
```

File: src/Approximations/overapproximate.jl (hunk 8 above)
Comment (schillic) [reply src/Approximations/overapproximate.jl:1695]: Note that these methods dispatching on `dir::Type{<:AbstractDirections{N}}` are never used. Should we rather remove them?

Comment (mforets) [reply src/Approximations/overapproximate.jl:1695]: i sometimes use them from outside

Comment (mforets) [reply src/Approximations/overapproximate.jl:1695]: consider it a feature :)

Comment (schillic) [reply src/Approximations/overapproximate.jl:1695]: Then maybe they should be tested. This one was broken.

Comment (SebastianGuadalupe) [reply src/Approximations/overapproximate.jl:1695]: Those were done to allow the user to do something like `overapproximate(X, Zonotope, OctDirections)` (without specifying the dimensionality)

Comment (schillic) [reply src/Approximations/overapproximate.jl:1695]: But they are not used in that case. You call these helper methods with the instantiated directions only from inside the library. That is why this error was not detected.

File: test/unit_overapproximate.jl [hunk 11]
```diff
@@ -349,6 +349,17 @@ for N in [Float64]
...
+    H = convert(HPolytope, B)
+    Z = Zonotope(N[0, 0, 0], N[1 0 0; 0 1 0; 0 0 1])
+    @test overapproximate(H, Zonotope, BoxDirections, algorithm="vrep") == Z
+    @test overapproximate(H, Zonotope, BoxDirections, algorithm="cpa") == Z
+    V = convert(VPolytope, B)
+    @test overapproximate(V, Zonotope, BoxDirections, algorithm="vrep") == Z
+    @test overapproximate(V, Zonotope, BoxDirections, algorithm="cpa") == Z
+    # test dispatch of internal functions
+    @test _overapproximate_zonotope_vrep(V, BoxDirections) == Z
+    @test _overapproximate_zonotope_cpa(V, BoxDirections) == Z
```
Comment (schillic) [inline test/unit_overapproximate.jl:362]:
```suggestion
    @test Approximations._overapproximate_zonotope_vrep(V, BoxDirections) == Z
    @test Approximations._overapproximate_zonotope_cpa(V, BoxDirections) == Z
```
(I guess, but we can check...)

File: test/unit_overapproximate.jl [hunk 12]
```diff
@@ -349,6 +349,17 @@ for N in [Float64]
...
+    B = BallInf(zeros(N, 3), N(1))
+    H = convert(HPolytope, B)
+    Z = Zonotope(N[0, 0, 0], N[1 0 0; 0 1 0; 0 0 1])
+    @test overapproximate(H, Zonotope, BoxDirections, algorithm="vrep") == Z
+    @test overapproximate(H, Zonotope, BoxDirections, algorithm="cpa") == Z
+    V = convert(VPolytope, B)
+    @test overapproximate(V, Zonotope, BoxDirections, algorithm="vrep") == Z
+    @test overapproximate(V, Zonotope, BoxDirections, algorithm="cpa") == Z
+    # test dispatch of internal functions
+    @test Approximations._overapproximate_zonotope_vrep(V, BoxDirections(3)) == Z
```
Comment (mforets) [inline test/unit_overapproximate.jl:361]: why not keep the test fo this function https://github.com/JuliaReach/LazySets.jl/pull/2298/files#diff-eb4fb67940b1678df4462dad2b59a25cR1691

Comment (SebastianGuadalupe) [reply test/unit_overapproximate.jl:361]: You are right, I will undo the change

**Comment by mforets** [comment]
looks good  to me.. merge? @schillic

**Pull request merged**